drivers/sensor/bosch/bmp180/bmp180.c

```diff
@@ -0,0 +1,499 @@
+/* Bosch BMP180 pressure sensor
+ *
+ * Copyright (c) 2024 , Chris Ruehl
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.mouser.hk/datasheet/2/783/BST-BMP180-DS000-1509579.pdf
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/pm/device.h>
+#include <math.h>
+
+#include "bmp180.h"
+
+LOG_MODULE_REGISTER(BMP180, CONFIG_SENSOR_LOG_LEVEL);
+
+static inline int bmp180_bus_check(const struct device *dev)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->check(&cfg->i2c);
+}
+
+static inline int bmp180_reg_read(const struct device *dev,
+				  uint8_t start, uint8_t *buf, int size)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->read(&cfg->i2c, start, buf, size);
+}
+
+static inline int bmp180_reg_write(const struct device *dev, uint8_t reg,
+				   uint8_t val)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->write(&cfg->i2c, reg, val);
+}
+
+int bmp180_reg_field_update(const struct device *dev,
+			    uint8_t reg,
+			    uint8_t mask,
+			    uint8_t val)
+{
+	int rc = 0;
+	uint8_t old_value, new_value;
+	const struct bmp180_config *cfg = dev->config;
+
+	rc = cfg->bus_io->read(&cfg->i2c, reg, &old_value, 1);
+	if (rc != 0) {
+		return rc;
+	}
+
+	new_value = (old_value & ~mask) | (val & mask);
+	if (new_value == old_value) {
+		return 0;
+	}
+
+	return cfg->bus_io->write(&cfg->i2c, reg, new_value);
+}
+
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+static int bmp180_attr_set_oversampling(const struct device *dev,
+					enum sensor_channel chan,
+					uint16_t val)
+{
+	struct bmp180_data *data = dev->data;
+
+	/* Value must be a positive value 0-3 */
+	if ((chan != SENSOR_CHAN_PRESS) || (val > 3)) {
+		return -EINVAL;
+	}
+
+	data->osr_pressure = val;
+
+	return 0;
+}
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+static int bmp180_attr_set(const struct device *dev,
+			   enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif /* CONFIG_PM_DEVICE */
+
+	switch (attr) {
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+	case SENSOR_ATTR_OVERSAMPLING:
+		ret = bmp180_attr_set_oversampling(dev, chan, val->val1);
+		break;
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+
+static int read_raw_temperature(const struct device *dev)
+{
+	int ret;
+	uint8_t ctrlreg;
+	struct reg16_t {
+		uint8_t msb;
+		uint8_t lsb;
+	} reg16;
```

Comment:
typedefed structs are usually not the way to define structures in zephyr. Also I am wondering if the concept of converting from big endian is such a stretch for someone reading driver code that this struct is needed to communicate/document it. I'd advise you to go with the buffer approach I initially suggested, however opinions of other reviewers also welcome on this.

Reply:
I change it to a uint8_t[] because you right if not packed it will be aligned and can cause unexpected sides.